### ISSUE TITLE
Revisit "narrowing" optimization rationale

### ISSUE THREAD

**Issue by lukego** [raptorjit/raptorjit#41]
The "narrowing numbers to integers" optimization in `lj_opt_narrow.c` includes an extremely well-written rationale for the approach taken with reference to the CPU microarchitectural details. (Inspiringly good piece of code commenting!) However, this was written circa 2010 in the Nehalem era which has since been superseded by Sandy Bridge and then Haswell and then Skylake. So it makes sense to revisit the design and see if it still makes sense or whether the assumptions that it is based on are no longer true. Could be an opportunity to improve performance by updating the optimization, or to simplify the code by removing optimizations that are no longer beneficial.

Let's use the [standard optimization resources](https://github.com/raptorjit/raptorjit#optimization-resources) for reference and especially this [illustration from AnandTech of the Haswell execution engine](http://www.anandtech.com/show/6355/intels-haswell-architecture/8) for reference:

![haswellexec](https://cloud.githubusercontent.com/assets/13791/24037049/8f767f94-0afc-11e7-9f59-c05790cc8f0f.png)

Here are some parts of the comments that jump out at me from a 2017 perspective:

> The total CPU execution bandwidth is the sum of the bandwidth of the FP and the integer units, because they execute in parallel. The FP units have an equal or higher bandwidth than the integer units. Not using them means losing execution bandwidth. Moving work away from them to the already quite busy integer units is a losing proposition.

This does not seem to be true of Haswell. Haswell has two execution ports that can do either float or integer operations (0,1) and two ports that are integer-only (5,6). So it seems like the opposite to the comment is true: integer operations have more bandwidth & and have to contend with float operations for use of execution resources.

> Always replacing an FP op with an integer op plus an overflow check is counter-productive on a current-generation super-scalar CPU. Although the overflow check branches are highly predictable, they will clog the execution port for the branch unit and tie up reorder buffers.

Is it still true that branch units and reorder entries are a scarce resource? Haswell has upgraded from one branch unit to two (ports 0,6) and it has a generous 192-entry reorder buffer.

So - summary - it seems like it would be worth reviewing the narrowing optimization and at least checking whether there is any complex code that could be retired as non-beneficial.

Relatedly: LuaJIT supports a diverse range of "number modes." I have experimented with these in the past, particularly trying to switch to ones that seem like they should help on Snabb's integer-heavy workloads, but I have never been able to observe a difference. So it seems possible that certain LuaJIT optimizations are only effective on older CPUs and can be removed in RaptorJIT. (I already removed non-canonical number modes in #5 but it could be worth revisiting whether I kept the right one i.e. the simplest of the fast options.)

**Comment by lukego** [comment]
... actually it looks like my use of `unifdef` in #5 was not extensive enough to remove the single-vs-dual number modes from RaptorJIT. This is lucky because now we can see which one makes the right performance/simplicity trade-off and remove the other!

**Comment by lukego** [comment]
I am slightly fascinated with the narrow optimization at the moment. I will do some tests to check how important it is in practice.

I suspect that it will prove to be beneficial: eliminating guard and conversion instructions, utilizing the integer-only ALUs (ports 5,6), and making other optimizations like `fold` more effective by letting them operate on integers instead of floats.

However, if it did prove to be unbeneficial then that would be a bit thrilling. The narrow optimization is a lot of tricky code that would be lovely to eliminate, and that may have the knock-on effect of being able to remove a bunch of integer-oriented optimizations from `fold`. I doubt it will pan out that way but feel obliged to check.

**Comment by lukego** [comment]
I tested `-O-narrow` on the Snabb CI. The [report](https://hydra.snabb.co/build/1093573/download/2/report.html) takes some practice to interpret but it shows a definite degradation of the `lukego-optimize` benchmark result probability distribution compared with the `master` reference one. Ballpark 5% slowdown on the `l2fwd` benchmarks (and no apparent impact on the `iperf` benchmark.) These benchmarks both do a lot of integer operations including with 16/32/64 bit FFI values.

So the good news is that `narrow` is beneficial in practice on Intel CPUs. The bad news is that the benefit may be fairly modest even though the code is quite complex.

I will look at some more optimizations and benchmarks.

**Comment by lukego** [comment]
Closing because the `narrow` optimization seems to be valuable in practice and not a candidate for removal.
